action/protocol/staking/handlers.go

```diff
@@ -551,18 +583,30 @@ func (p *Protocol) fetchBucket(
 	index uint64,
 	checkOwner bool,
 	allowSelfStaking bool,
-) (*VoteBucket, error) {
+) (*VoteBucket, *fetchError) {
 	actionCtx := protocol.MustGetActionCtx(ctx)
 	bucket, err := getBucket(sr, index)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to fetch bucket by index %d", index)
+		fetchErr := &fetchError{
+			err:           errors.Wrapf(err, "failed to fetch bucket by index %d", index),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketIndex,
+		}
+		return nil, fetchErr
 	}
 	if checkOwner && !address.Equal(bucket.Owner, actionCtx.Caller) {
-		return nil, fmt.Errorf("bucket owner does not match action caller, bucket owner %s, action caller %s",
-			bucket.Owner.String(), actionCtx.Caller.String())
+		fetchErr := &fetchError{
+			err: fmt.Errorf("bucket owner does not match action caller, bucket owner %s, action caller %s",
+				bucket.Owner.String(), actionCtx.Caller.String()),
+			failureStatus: iotextypes.ReceiptStatus_ErrUnauthorizedOperator,
+		}
+		return nil, fetchErr
 	}
 	if !allowSelfStaking && p.inMemCandidates.ContainsSelfStakingBucket(index) {
-		return nil, errors.New("self staking bucket cannot be processed")
+		fetchErr := &fetchError{
+			err:           errors.New("self staking bucket cannot be processed"),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketType,
+		}
```

Comment:
this is also first type error

Reply:
can we return error with errors.Wrap(Errxxx,"xxxx"),so I can use errors.Cause(err) to check when write unit test

Reply:
@lzxm160 This would be a nonerr case with a failure receipt. You may need to check the receipt status instead of error.